core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1539,6 +1539,13 @@ abstract class RDD[T: ClassTag](
     // NOTE: we use a global lock here due to complexities downstream with ensuring
     // children RDD partitions point to the correct parent partitions. In the future
     // we should revisit this consideration.
+    if (doCheckpointCalled) {
+      throw new SparkException("Because job has been executed on this RDD, checkpoint won't work")
+    }
+    if (storageLevel == StorageLevel.NONE) {
+      logWarning("Call checkpoint on unpersisted RDD, it will cause RDD recomputation" +
+        " when saving checkpoint files.")
+    }
```

Comment:
Two issues with this :

a) It is possible for cache/persist to be called after checkpoint() - and behave the same as cache/persist called before checkpoint()
b) A narrow dependency parent could have been persisted or checkpoint'ed - causing this checkpoint to not be expensive : so the message and log level is a bit too severe.